### PR TITLE
feat(generator): support services with location-dependent endpoints

### LINKED ISSUE #7976: Cloud Document AI API: `documentai.googleapis.com`
- [x] Make the quickstart work, and run it in our ci builds.
- [ ] Add a helper function for setting the EndpointOption.

### PR DESCRIPTION
**Pull request by devbww**
Add generator support for services with location-dependent default
endpoints.  A new `MakeServiceConnection()` overload is generated with
a required `location` parameter, which is used to build the default
`EndpointOption` of `"<location>-<normal-service-endpoint>"` (for
example, "us-documentai.googleapis.com" in the Cloud Document AI API).

For backwards compatibility in already-released libraries, the old
`MakeServiceConnection()` signature is still available, but it, as
before, requires setting `EndpointOption` and `AuthorityOption` to
get something that works.

Fixes #7976 and #8704.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9625)

### CODE REVIEW

File: generator/internal/codegen_utils.cc
Comment (dbolduc) [inline generator/internal/codegen_utils.cc:126]: can you add a unit test to `codegen_utils_test.cc`?

**Review by dbolduc — CHANGES_REQUESTED** [review]
I like this. I just think we need to incorporate it into `test.proto` so we can unit test the functionality of `Make*Connection(std::string location const&, Options options)`. (Or alternatively, add unit tests to one of the actual libraries like `documentai`. This is not how we have done things in the past though)

File: generator/internal/connection_generator.cc
Comment (coryan) [inline generator/internal/connection_generator.cc:264]: I am not sure it is always useless, some services have global and regional resources, with global and regional endpoints.  Maybe "Unless the service also offers a global endpoint, the default value of the `EndpointOption` may be useless ..." ?

**Review by coryan — COMMENTED** [review]
Generally, I am inclined to merge these changes.  I have some questions for the future:

I wonder what other languages did for these regional endpoints.

What happens if you configure a `us-foo.googleapis.com` endpoint and then try to use resources in the `eu` region?

Should we (maybe) have a custom function that maps resource names to endpoints and then to different channels / stubs?  Or maybe that is too complicated and we should only do this if there is enough demand?

**Comment by devbww** [comment]
> Darren: I just think we need to incorporate it into `test.proto` so we can unit test the functionality of `Make*Connection(std::string location const&, Options options)`.

The only current way to do that seems to be to update `golden_config.textproto:service`, but that would then not exercise the normal case.  My thought was to leave this to the quickstart for the location-dependent services, which has to wait for the next release and renovation, but which is on my to-do list.

> Carlos: I wonder what other languages did for these regional endpoints.

I did look at some, and they did something similar.  The only one I can seem to re-discover at the moment is [python document-ai example](http://cloud/document-ai/docs/libraries#client-libraries-usage-python), where we see stuff like:
```
    # You must set the api_endpoint if you use a location other than 'us', e.g.:
    opts = ClientOptions(api_endpoint=f"{location}-documentai.googleapis.com")
```

> Carlos: What happens if you configure a `us-foo.googleapis.com` endpoint and then try to use resources in the `eu` region?

¯\\_(ツ)_/¯

> Carlos: Should we (maybe) have a custom function that maps resource names to endpoints and then to different channels / stubs? Or maybe that is too complicated and we should only do this if there is enough demand?

I reckon that does sound too complicated, at least until we have some experience/feedback on the simple tactic.

File: generator/internal/option_defaults_generator.cc
Comment (dbolduc) [reply generator/internal/option_defaults_generator.cc:136]: errr.... fine. But if the global endpoint has utility, customers would have to set it using `EndpointOption` or the endpoint env var. (vs. `Make*FooConnection("")`)

Comment (devbww) [reply generator/internal/option_defaults_generator.cc:136]: **If** a service's global endpoint has utility, I'd suggest that we're probably best off with an explicit "endpoint style" for that case.

**Comment by dbolduc** [comment]
> The only current way to do that seems to be to update `golden_config.textproto:service`, but that would then not exercise the normal case.

oops I told you the wrong file to update. We *could* add another service to the `.textproto` file. But I do not think it is worth the bloat.

> My thought was to leave this to the quickstart for the location-dependent services, which has to wait for the next release and renovation, but which is on my to-do list.

1. We should open a new issue to update the quickstarts for location-dependent services.

2. Quickstarts are integration tests, and if we can add unit tests we probably should. I kind of like the idea of unit testing our actual generated libraries. We can discuss if there's any merit to that idea in the team meeting.

**Review by dbolduc — APPROVED** [review]

**Review by coryan — APPROVED** [review]

**Pull request merged**
